dotcom-rendering/scripts/islands/descriptions.mjs

```diff
@@ -0,0 +1,146 @@
+// @ts-check
+
+import { readdir, readFile, writeFile } from 'node:fs/promises';
+import { resolve } from 'node:path';
+import { fileURLToPath } from 'node:url';
+import rehypeAutolinkHeadings from 'rehype-autolink-headings';
+import rehypeSlug from 'rehype-slug';
+import rehypeStringify from 'rehype-stringify';
+import remarkParse from 'remark-parse';
+import remarkRehype from 'remark-rehype';
+import { unified } from 'unified';
+import * as zod from 'zod';
+
+/** @type {(name: string) => RegExp} */
+export const get_JSDoc = (name) =>
+	new RegExp(
+		'\n/\\*\\*\n?' +
+			'((?: \\*.*\n)+)' +
+			' \\*\\/' +
+			`\nexport const ${name} =`,
+		'm',
+	);
+
+const components_directory = resolve(
+	fileURLToPath(import.meta.url),
+	'../../../src/web/components',
+);
+
+const PREFIX = 'window.chartData = ';
+
+const chunk = zod.object({
+	label: zod.string(),
+	isAsset: zod.boolean(),
+	statSize: zod.number(),
+	parsedSize: zod.number(),
+	gzipSize: zod.number(),
+});
+
+const bundle_report = await readFile(
+	resolve(
+		fileURLToPath(import.meta.url),
+		'../../../dist/browser.modern-bundles.html',
+	),
+	'utf-8',
+)
+	.then((file) =>
+		file
+			.split('\n')
+			.find((line) => line.trim().startsWith(PREFIX))
+			?.trim(),
+	)
+	.then((bundle_data) =>
+		zod
+			.array(chunk)
+			.parse(
+				JSON.parse(
+					bundle_data?.slice(PREFIX.length, -1).trim() ?? '[]',
+				),
+			),
+	);
+
+const processor = unified()
+	.use(remarkParse)
+	.use(remarkRehype)
+	.use(rehypeSlug)
+	.use(rehypeAutolinkHeadings, { behavior: 'wrap' })
+	.use(rehypeStringify);
+
+const islands = await readdir(components_directory)
+	.then((components) =>
+		components.filter((component) => component.endsWith('.importable.tsx')),
+	)
+	.then((files) =>
+		Promise.all(
+			files.map(async (file) => ({
+				file,
+				content: await readFile(
+					resolve(components_directory, file),
+					'utf-8',
+				),
+			})),
+		),
+	)
+	.then((files_and_content) =>
+		Promise.all(
+			files_and_content.map(async ({ file, content }) => {
+				const name = file.replace('.importable.tsx', '');
+				const { gzipSize = 0, parsedSize = 0 } =
+					bundle_report.find(({ label }) =>
+						label.startsWith(name + '-importable.'),
+					) ?? {};
+
+				const REGEX = get_JSDoc(name);
+
+				const matched = content.match(REGEX);
+
+				const { index, match } = matched
+					? { match: matched[1], index: matched.index ?? -1 }
+					: { match: undefined, index: -1 };
+
+				const lines = match
+					? '#' +
+					  [
+							content.slice(0, index).split('\n').length + 1,
+							content.slice(0, index + match.length).split('\n')
+								.length + 2,
+					  ]
+							.map((line_count) => `L${line_count}`)
+							.join('-')
+					: '';
+
+				const description =
+					matched?.[1]
+						?.split('\n')
+						.map((jsdoc_line) => jsdoc_line.slice(3))
+						.join('\n') ?? `# ${name} \n No description yet… 😢`;
```

Comment:
> Encourage good documentation

Is this implying we expect a `description`? Some Islands are self describing, so maybe we're just creating work here for engineers - and re-describing things?

Reply:
Just adding my two cents as a non-engineer and a newbie to GNM that descriptions expanding on their names would certainly be useful to me personally (and I do plan to encourage my fellow non-devs to utilise such documentation)

Reply:
>  Explaining why an Island cannot be server-side rendered s

That sounds great - should we do that with our current descriptions?

Reply:
> >  Explaining why an Island cannot be server-side rendered s
> 
> 
> 
> That sounds great - should we do that with our current descriptions?

I would like that an plan to help with this effort!